libkbfs/bserver_remote.go

```diff
@@ -405,25 +380,39 @@ func (b *BlockServerRemote) batchDowngradeReferences(ctx context.Context,
 			}
 			nonces[BlockRefNonce(ref.Ref.Nonce)] = ref.LiveCount
 		}
-
-		//figure out the not-yet-deleted references
-		oldNotDoneLen := len(notDone)
+		// update the list of references to downgrade
 		notDone = b.getNotDone(contexts, doneRefs)
-		if !madeProgress {
-			madeProgress = len(notDone) < oldNotDoneLen
-		}
-		if !madeProgress {
-			return doneRefs,
-				errors.New("Made no progress after a batch downgrade call")
-		}
 
 		//if context is cancelled, return immediately
 		select {
 		case <-ctx.Done():
-			return doneRefs, ctx.Err()
+			finalError = ctx.Err()
+			return nil
 		default:
 		}
 
+		// check whether to backoff and retry
+		if err != nil {
+			// if error is of type throttle, retry
+			if _, ok := err.(BServerErrorThrottle); ok {
+				return err
+			}
+			// non-throttle error, do not retry here
+			finalError = err
+		}
+		return nil
+	}, backoff.NewExponentialBackOff())
```

Comment:
Can we put a `MaxElapsedTime` on this, to make sure we don't loop forever on systems where there is no context timeout?